### ISSUE TITLE
[discussion] Modernization and ES2015+

### ISSUE THREAD

**Issue by jayands** [themasch/node-ebml#35]
I noticed that the latest commit is a chore to modernize the code. Only, it only goes so far, as if you're afraid to actually write code *for* modern Node because of worrying of breaking Node 4 support or something. [Babel][babel] and/or [Rollup][rollup] can alleviate these issues, especially [`babel-preset-env`][env], which can be configured to only transpile the code, as needed, with a few [^1] JSON options:

```JSONC
{
  "presets": ["stage-0",
    ["env", {
      "targets": {
        // "browsers": ["last 2 versions", "> 1%"], // see https://github.com/ai/browserslist
        "node": "current"
      }
    }]
  ]
}
```
The setup for [Babel 7][@babel], however, is slightly different: everything is under the `@babel` namespace, and they seem to have deprecated `@babel/preset-stage-0`, as well as stages 1 and 2 (Stage 3 is considered by the [TC39][es-processes] in candidate stage and Stage 4 is what you get in the current release, so like arrow functions, async/await, etc.). The setup for that JSON file is a bit different:

```JSONC
{
  "plugins": [
    // see https://github.com/babel/babel/tree/master/packages/babel-preset-stage-0 for a list of plugins
  ],
  "presets": [
    "@babel/stage-3",
   ["@babel/env", { /* same options as above */ } ]
  ]
}
```
On the other hand, I can see reasons *against* adding in a transpiler dependency, namely that it adds in otherwise unnecessary complexity to the build process.

I'd like to get your opinion; I've forked but not pulled to start working on it, plus I left some comments on the last update, and would love your feedback either way.

[^1]: [`babel-preset-stage-0`][stage-0] is a preset bundle that adds in functionality for all the JS proposals

[env]: https://babeljs.io/docs/en/babel-preset-env
[stage-0]: https://babeljs.io/docs/en/babel-preset-stage-0
[rollup]: https://rollupjs.org
[babel]: https://babeljs.io
[@babel]: https://babeljs.io/docs/en/next/index.html
[es-processes]: http://2ality.com/2015/11/tc39-process.html

**Comment by themasch** [comment]
I thought about using something like that, too but for browser support. 
Node 4.x is out of maintenance since may and I guess we shouldn't really care to support all of this old version. 

I don't really want to go there and release compiled sources into npm for use in node. But I'm thinking about doing that for browser use. 

But its totally true that the last code transformation could use some work. I just did a really quick and dirty move with a lot of searches & replace but did rewrite every part with the most modern constructs. 

I also saw your comments. I hope to find some time next week (finally vacation) to invest some time in node-ebml. While I don't exactly follow your reasoning with all the purposed changes, I'll definitely rewrite some parts for the library to be more readable. 

tl;dr: I don't think we should or use babel for compiling, but I'm looking into ways to make the life for those who use it in the browser easier.

**Comment by jayands** [comment]
Alright. That's fair. You can see what's actually supported in Node versions, however, at https://node.green.
One thing to note, however: there is no actual reason _not_ to use arrow functions, *especially* in the test suite, if you're only supporting modern-ish Node.js. They can be a bit confusing at first, but they ultimately clean up scope if they are consistently used in place of anonymous functions

**Comment by jayands** [comment]
Alright, but if you're planning on supporting ES2015+, you might want to take a look at https://node.green as well as adding in an ["engines"][pkg.engines] key to the `package.json`. This could be supplemented by paying attention to the [Node release schedule][node-release]; currently, node 6.x is on the way out (next April), and the odd versions between 6 and 10 aren't being supported at all, with cutting-edge 11 being the only odd-numbered version to worry about in terms of support/bug fixing.

[pkg.engines]: https://docs.npmjs.com/files/package.json#engines
[node-release]: https://github.com/nodejs/Release
[mdn-arrow]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

One thing I do want to touch on though: [arrow functions][mdn-arrow]. Basically, there are very few reasons *not* to use them, especially with the testing environment. They clean up the scope of your environment considerably when used consistently, and a lot of the functions that need to mutate `this`, as functions, could probably be rewritten and have `this` passed to them as variables instead (unless they're in the class prototype, but chances are that they don't have the `function` keyword attached, etc.)
